execution/function/function.go

```diff
@@ -1,247 +1,482 @@
 // Copyright (c) The Thanos Community Authors.
 // Licensed under the Apache License 2.0.
 
-package scan
+package function
 
 import (
+	"context"
 	"fmt"
 	"math"
+	"sort"
+	"sync"
 
 	"github.com/efficientgo/core/errors"
 	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/value"
 	"github.com/prometheus/prometheus/promql"
 	"github.com/prometheus/prometheus/promql/parser"
 
+	"github.com/thanos-community/promql-engine/execution/model"
 	"github.com/thanos-community/promql-engine/execution/parse"
 )
 
+// functionOperator returns []model.StepVector after processing input with desired function.
+type functionOperator struct {
+	funcExpr    *parser.Call
+	series      []labels.Labels
+	once        sync.Once
+	vectorIndex int
+	nextOps     []model.VectorOperator
+
+	call FunctionCall
+
+	scalarInput  [][]model.StepVector
+	scalarPoints []float64
+}
+
+func NewfunctionOperator(funcExpr *parser.Call, call FunctionCall, nextOps []model.VectorOperator) (model.VectorOperator, error) {
+	f := &functionOperator{
+		nextOps:      nextOps,
+		call:         call,
+		funcExpr:     funcExpr,
+		vectorIndex:  0,
+		scalarInput:  make([][]model.StepVector, len(nextOps)-1),
+		scalarPoints: make([]float64, len(nextOps)-1),
+	}
+
+	for i := range funcExpr.Args {
+		if funcExpr.Args[i].Type() == parser.ValueTypeVector {
+			f.vectorIndex = i
+			break
+		}
+	}
+
+	// Check selector type.
+	// TODO(saswatamcode): Add support for string and matrix.
+	switch funcExpr.Args[f.vectorIndex].Type() {
+	case parser.ValueTypeVector, parser.ValueTypeScalar:
+		return f, nil
+	default:
+		return nil, errors.Wrapf(parse.ErrNotImplemented, "got %s:", funcExpr.String())
+	}
+}
+
+func (o *functionOperator) Explain() (me string, next []model.VectorOperator) {
+	return fmt.Sprintf("[*functionOperator] %v(%v)", o.funcExpr.Func.Name, o.funcExpr.Args), o.nextOps
+}
+
+func (o *functionOperator) Series(ctx context.Context) ([]labels.Labels, error) {
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	return o.series, nil
+}
+
+func (o *functionOperator) GetPool() *model.VectorPool {
+	return o.nextOps[o.vectorIndex].GetPool()
+}
+
+func (o *functionOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	// Process non-variadic single/multi-arg instant vector and scalar input functions.
+	// Call next on vector input.
+	input, err := o.nextOps[o.vectorIndex].Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(input) == 0 {
+		return nil, nil
+	}
+
+	// Call next on all scalar inputs.
+	if len(o.scalarInput) != 0 {
+		s := 0
+		for i := range o.nextOps {
+			if i == o.vectorIndex {
+				continue
+			}
+
+			si, err := o.nextOps[i].Next(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			o.scalarInput[s] = si
+			defer o.nextOps[i].GetPool().PutVectors(o.scalarInput[s])
```

Comment:
I think this will miss the recycling of the vector operator samples

Reply:
Have added defer `PutStepVector` statements to release samples. Both for `scalarInput` as well as `input`